src/connection/mqtt/mod.rs

```diff
@@ -0,0 +1,844 @@
+                        debug!("Purging properties");
+
+                        self.purge_properties(device, &publish.payload).await?;
+                    } else {
+                        let client_id = CLIENT_ID.get_or_init(|| self.client_id().to_string());
+                        let ParsedTopic { interface, path } =
+                            ParsedTopic::try_parse(client_id, &publish.topic)?;
+
+                        return Ok(ReceivedEvent {
+                            interface: interface.to_string(),
+                            path: path.to_string(),
+                            payload: publish.payload,
+                        });
+                    }
+                }
+                packet => {
+                    trace!("packet received {packet:?}");
+                }
+            }
+        }
+    }
+
+    fn deserialize_individual(
+        &self,
+        mapping: MappingRef<'_, &Interface>,
+        payload: &Self::Payload,
+    ) -> Result<(AstarteType, Option<Timestamp>), crate::Error> {
+        payload::deserialize_individual(mapping, payload).map_err(|err| err.into())
+    }
+
+    fn deserialize_object(
+        &self,
+        object: ObjectRef,
+        path: &MappingPath<'_>,
+        payload: &Self::Payload,
+    ) -> Result<(HashMap<String, AstarteType>, Option<Timestamp>), crate::Error> {
+        payload::deserialize_object(object, path, payload).map_err(|err| err.into())
+    }
+
+    async fn send_individual(
+        &self,
+        validated: ValidatedIndividual<'_>,
+    ) -> Result<(), crate::Error> {
+        let buf = payload::serialize_individual(validated.data(), validated.timestamp())?;
+
+        self.send(
+            validated.mapping().interface(),
+            validated.path().as_str(),
+            validated.mapping().reliability().into(),
+            buf,
+        )
+        .await
+    }
+
+    async fn send_object(&self, validated: ValidatedObject<'_>) -> Result<(), crate::Error> {
+        let buf = payload::serialize_object(validated.data(), validated.timestamp())?;
+
+        self.send(
+            validated.object().interface,
+            validated.path().as_str(),
+            validated.object().reliability().into(),
+            buf,
+        )
+        .await
+    }
+
+    async fn added_interface<S>(
+        &self,
+        device: &SharedDevice<S>,
+        added_interface: &str,
+    ) -> Result<(), crate::Error>
+    where
+        S: Send + Sync,
+    {
+        let interfaces = device.interfaces.read().await;
+        let interface_ownership = interfaces
+            .get(added_interface)
+            .ok_or(InterfaceError::InterfaceNotFound {
+                name: added_interface.to_string(),
+            })?
+            .ownership();
+        let introspection_string = interfaces.get_introspection_string();
+
+        if interface_ownership == Ownership::Server {
+            self.subscribe(added_interface).await?
+        }
```

Comment:
This is intended since we want to avoid keeping the read lock created at line 466 between await points when the network is involved.
This is why we avoided moving around the variables.